[ds-binary-tree]

pathfinder177 opened the pull request:
* adding binary tree

Eugen-Kuzovlev committed:
[ds-binary-tree] added tree_create method

Eugen-Kuzovlev committed:
[ds-binary-tree] added pre,in,post order traversal methods

Eugen-Kuzovlev committed:
[ds-binary-tree] added tree_create_node method

Eugen-Kuzovlev committed:
[ds-binary-tree] added interface(function prototypes)

Eugen-Kuzovlev committed:
[ds-binary-tree] added dfs search

Eugen-Kuzovlev committed:
[ds-binary-tree] in tree_create_node added malloc check

Eugen-Kuzovlev committed:
[ds-binary-tree] fixed comparison intree_create_node; added bfs draft

Eugen-Kuzovlev committed:
[ds-binary-tree] added bfs draft using queue

Eugen-Kuzovlev committed:
[ds-binary-tree] added working bfs draft using queue

Eugen-Kuzovlev committed:
[ds-binary-tree] struct Node: put in h file

Eugen-Kuzovlev committed:
[ds-binary-tree] bfs: fixed while condition

Eugen-Kuzovlev committed:
[ds-binary-tree] added tree_get_node_val method

Eugen-Kuzovlev committed:
[ds-binary-tree] added tree_set_node_val method

Eugen-Kuzovlev committed:
[ds-binary-tree] added tree_get for left, right children methods

Eugen-Kuzovlev committed:
[ds-binary-tree] tree_get methods return NULL instead of FAILURE

Eugen-Kuzovlev committed:
[ds-binary-tree] add tree_insert left and right methods

Eugen-Kuzovlev committed:
[ds-binary-tree] added insertion etc by methods, not hands. delete_node in pseudocode

Eugen-Kuzovlev committed:
[ds-binary-tree] added tree_search_parent method

Eugen-Kuzovlev committed:
[ds-binary-tree] tree_delete_node: handled root, leaf and only left or right node

Eugen-Kuzovlev committed:
[ds-binary-tree] tree_delete_node: added as in BST(see Stivens Chapter 10)

Yanpas commented on bfs_queue.h:
```diff
@@ -0,0 +1,6 @@
+
+struct Node* items[SIZE];
+void enQueue(struct Node *);
+struct Node* deQueue();
```
is bfs related to bin tree?

pathfinder177 replied:
yes sir, i believe so

Yanpas commented on binary_tree.c:
```diff
@@ -0,0 +1,410 @@
+    }
+}
+
+struct Node* tree_search_last_right_vertice(struct Node* left_child) {
```
I guess it's not a public interface of a tree. So if it's meant to be used only inside binary_tree.c we usually do not put such functions to headers, and in source files mark them static

Yanpas commented on binary_tree.c:
```diff
@@ -0,0 +1,410 @@
+    }
+}
+
+void tree_delete_node(Tree* tree, struct Node* self) {
```
the naming is confusing, please rename tree to self. Also I'd suggest such interface to avoid parent_lookup:
`bool tree_erase(Tree* self, int value)`

Yanpas commented on binary_tree.c:
```diff
@@ -0,0 +1,410 @@
+
+void tree_delete_node(Tree* tree, struct Node* self) {
+    //if duplicate is encountered: delete first found node
+    struct Node* parent = tree_search_parent(tree, self);
```
this is log_n operation, please avoid it

Yanpas commented on binary_tree.c:
```diff
@@ -0,0 +1,410 @@
+    }
+    else if(parent->l_node && parent->r_node == NULL) {
+        if(tree->root == self) {
+            tree->root = self->l_node;
```
tree root shouldn'r be affected at all, imagine you have a tree:
```
    5
  3   8
2 4 7 9
     6
```
and you want to remove node 7. root shouldn't be affected at all. Later on, please write unit tests and tree_print

pathfinder177 replied:
I should have written comment that in this incorrect method until `else` I handled root with or without child nodes. In addition since the method is not recursive and has needless complexity

Yanpas commented on binary_tree.c:
```diff
@@ -0,0 +1,410 @@
+struct Node* tree_create_node(int);
+void tree_delete_node(Tree*, struct Node*);
+
+int tree_get_node_val(struct Node*);
```
Do you really need these functions?

pathfinder177 replied:
I copied it from the very basic tree interface.
See no problems if it remains here.
I tested it as other functions
Wdyt?

Yanpas replied:
Well, instead of calling this function we can always do `node->data`, and it makes no difference.

Yanpas commented on binary_tree.c:
```diff
@@ -0,0 +1,410 @@
+
+
+Tree tree_create(); // returns empty binary tree
+// void tree_free(Tree* self); // deallocates tree
```
well, this one is important, you can use call stack if you need for DFS

Yanpas commented on binary_tree.c:
```diff
@@ -0,0 +1,410 @@
+    return n;
+}
+
+struct Node* tree_dfs_search(struct Node* self, int value) {
```
well, DFS is a traverse method, you don't need to search with DFS in binary tree, since it's already sorted. There's only one search algorithm possible, it's comparing if current node is less, greater or equal to the vlue being searched

pathfinder177 replied:
Shall I add sort method then?

Yanpas replied:
No. The tree always remains sorted

Yanpas commented on binary_tree.c:
```diff
@@ -0,0 +1,410 @@
+}
+
+void tree_delete_node(Tree* tree, struct Node* self) {
+    //if duplicate is encountered: delete first found node
```
duplicates can't exist in bin tree: there can be only one node with particular value

Eugen-Kuzovlev committed:
[ds-binary-tree] added tree_print function

[ds-binary-tree] added tree_free function

[ds-binary-tree] fixed incorrect traverses

[ds-binary-tree] added static to inner search functions

Eugen-Kuzovlev committed:
[ds-binary-tree] deleted tree_get_node_val function

Eugen-Kuzovlev committed:
[ds-binary-tree] added tree_rotate_left

Eugen-Kuzovlev committed:
[ds-binary-tree] added tree_rotate_right

Yanpas commented on binary_tree.c:
```diff
@@ -128,6 +130,60 @@ static void tree_rotate_left(Tree* self, int value) {
     }
 }
 
+static void tree_rotate_right(Tree* self, int value) {
```
why do you accept value? What if this value won't be found? I suggest you to accept `Tree*` and `Node*` which will be rotated to the right (it's assumed that it has left child). Also don't forget to check If the node is a root, in this case tree root should be updated

Yanpas commented on binary_tree.c:
```diff
@@ -128,6 +130,60 @@ static void tree_rotate_left(Tree* self, int value) {
+            if (parent->r_node && parent->r_node->data == value) {
+                ex_subtree_root = parent->r_node;
+
+                if(ex_subtree_root->l_node) {
```
this is a code duplication, you should avoid it `DRY` principle

pathfinder177 replied:
Here I agree, it can be expressed as an apart function for example
There is `WET` principle as well altough DRY indeed is better to apply here.

Yanpas commented on binary_tree.c:
```diff
@@ -359,21 +415,22 @@ int main() {
-    tree_insert_left(rotated_left_tree.root, 3);
-    tree_insert_left(rotated_left_tree.root->l_node, 1);
-    tree_insert_right(rotated_left_tree.root->l_node, 2);
+    tree_insert_left(rotated_tree.root, 3);
```
No! The tree is always sorted, you caný insert ot the left or right. You always start from the root and end up inserting a node as a **leaf**

Eugen-Kuzovlev committed:
[ds-bst] started bst transformation

[ds-bst] added tree_insert_node function

[ds-bst] added tree_find function

[ds-bst] added tree_find_parent function

Eugen-Kuzovlev committed:
[ds-bst] tree_erase_node draft

Eugen-Kuzovlev committed:
[ds-bst] tree_erase_node: handled leaf and if root has left node cases

Eugen-Kuzovlev committed:
[ds-bst] tree_erase_node: handled if root has only right case

Eugen-Kuzovlev committed:
[ds-bst] tree_erase_node: handled if value is root and root has left and right

Eugen-Kuzovlev committed:
[ds-bst] tree_erase_node: draft: erase_node refactored

Eugen-Kuzovlev committed:
[ds-bst] tree_erase_node: tested on nodes with and without child

Eugen-Kuzovlev committed:
[ds-bst] tree_erase_node: could not delete root node properly

Eugen-Kuzovlev committed:
[ds-bst] tree_erase_node: could not delete root node properly

Eugen-Kuzovlev committed:
[ds-bst] tree_erase_node: added working function

Eugen-Kuzovlev committed:
[ds-bst] replaced int to [0] in tree initialisation for

[ds-bst] added tree_rotation: left, right

Yanpas commented:
How about writing tree_equal function? :)
